[CBRD-25709] Trigger DEFERRED 구문에서 재귀 구조 발생 시 무한 루프 발생(Maximum trigger depth 미작동)

kangmin5505 opened the pull request:
http://jira.cubrid.org/browse/CBRD-25709

Purpose
- deferred trigger 에서 재귀 구조가 발생했을 시, maximum trigger depth 에러가 발생

Implementation
- BFS(Breath-First Search) 와 같이 한 layer 를 하나의 depth 로 묶어서,  before and after trigger 와 동일하게 Maximum trigger depth 를 넘었을 때 에러를 발생시킴

kangmin5505 committed:
try to test using directly free_state instead of using tr_finish

kangmin5505 committed:
try to apply local tr_curr_step

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25709

kangmin5505 committed:
Use tr_curr_step in +1 tr_Current_depth

kangmin5505 committed:
Add an annotation

ctshim commented on src/object/trigger_manager.c:
```diff
@@ -6009,85 +5901,87 @@ int
 	    {
-	      if (its_deleted (t->target))
+	      ++tr_curr_step;
+	      if (compare_recursion_levels (tr_curr_step, tr_Maximum_depth) > 0)
```
tr_Maximum_depth 변수는 실제 depth를 의미하는 것인데, tr_curr_step은 dept가 아니라 대상이 되는 숫자라서 비교가 적절하지 않은 것 같습니다.
가령 어떤 update 트리거가 있을 때 update된 레코드가 1000개라면 저 값은 1000이 될 수 있어야 합니다.

  ```
create table foo(n int);
  insert into foo select rownum from db_class a, db_class b limit 100;  
  CREATE TRIGGER foo_trigger5 DEFERRED UPDATE ON foo IF obj.n >= 1 EXECUTE PRINT 'foo_trigger  1';
  update foo set n = n+1;
```
위와 같이 하면 tr_curr_step는 100 된다는...

kangmin5505 replied:
피드백 감사드립니다.
네 말씀해 주신 부분이 맞습니다.
그 부분까지 고려해서 다시 요청드리겠습니다.

kangmin5505 replied:
말씀해 주신 부분을 고려하여 수정하였습니다.
요약하자면, while 문을 시작할 때를 기준으로 head 부터 tail 까지를 한 step 으로 묶어서, depth 와 동일하게 32 step(묶음)을 넘어가면 에러를 발생시키는 방법으로 구현했습니다.
더 자세한 내용은 회의 때 말씀드리도록 하겠습니다.

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25709

kangmin5505 committed:
Implement BFS structure for deferred trigger.

Applied a Breadth-First Search(BFS) structure to deferred triggers,
distinguishing them from before and after triggers, which follow a
Depth-First Search(DFS) structure.

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25709

ctshim commented on src/object/trigger_manager.c:
```diff
@@ -6025,59 +6024,77 @@ tr_execute_deferred_activities (DB_OBJECT * trigger_object, DB_OBJECT * target)
+		    }
 		}
 	    }
+	  while (t != tail && !error && (t = next));
```
while()의 조건문 체크 과정에서 조건으로 사용되는 변수를 변경하는 치환문이 들어 있는 구조는 
추후 코드를 다시 보게 될 때 혼동을 주거나 오인하게 될 수 있을 것 같습니다.

여기는 `while (t != tail && !error); `만 남겨두고  
6504라인에  `next = t;`
6057라인의 내용 앞쪽에 `t = next;`를 추가하는 것은 어떤가요?

```
          tail = c->tail;	/* range [head, tail] */
         next =t;

	  do
	    {
	          t = next;
                 next = t->next;
                  ...
            }
            while (t != tail && !error)


```

kangmin5505 replied:
코드 리뷰 감사합니다.
기존 코드에서는 t에 next 를 대입 후 NULL 인지 확인하는 로직이 있어서 조언해 주신 코드에서 while 조건문에 next 를 확인하는 로직이 필요할 것 같습니다.
다음과 같이 수정하도록 하겠습니다.

```
          tail = c->tail;	/* range [head, tail] */
         next =t;

	  do
	    {
	          t = next;
                 next = t->next;
                  ...
            }
            while (t != tail && !error && next != NULL)
```

kangmin5505 replied:
코드 리뷰 감사합니다.
말씀해 주신 조건을 for loop 로 작성하는 것이 더 깔끔할 것이라 판단하여, for loop 로 작성 후 다시 요청드리겠습니다.

ctshim commented on src/object/trigger_manager.c:
```diff
@@ -6025,59 +6024,77 @@ tr_execute_deferred_activities (DB_OBJECT * trigger_object, DB_OBJECT * target)
+		      status = execute_activity (trigger, TR_TIME_DEFERRED, t->target, NULL, &rejected);
+
+		      /* execute_activity() maybe include trigger and change the next pointer. we need get it again. */
+		      next = t->next;
```
assert(next == NULL ||  next == t->next);
next = t->next;

kangmin5505 replied:
코드 리뷰 감사합니다.
춘택님께서는 execute_activity 에 의해 기존의 next 를 잃어버리는 상황이 희박하거나 불가능하다고 하더라도, assert 코드를 작성하여 좀 더 명확하게 표현하는 것이 더 나은 방법이라고 생각하시는 것이 맞을까요?

kangmin5505 committed:
refactor while and do-while loops to for loops for improved readability
and consistency.

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25709

kangmin5505 committed:
codestyle

kangmin5505 committed:
Add assertion to ensure consistency of next pointer after trigger
execution

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25709

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25709

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25709

kangmin5505 commented:
/run all
